### PR TITLE
Prompt Decoding after truncation should not skip special tokens

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?

Quick fix for #3644. Skipping special tokens gets rid of the `<|im_start|>` and `<|im_end|>` chat templating and this text subsequently passed to vllm leads to malformed generations.

However, the way the prompt is currently truncated has a few other issues. For one it gets rid of the very first <|im_start|> if we exceed the `max_prompt_length` configuration. I'll work on a subsequent PR to truncate and reformat the prompt post truncation cleanly and ensure that the chat template remains in-tact post truncation.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by LeonEricsson** [comment]
Do you plan to split the `skip_special_tokens=False` fix into a separate PR? I think it makes sense to separate them – this way, we can get the fix merged as soon as possible.

**Comment by pramodith** [comment]
Cool, I'll have a new PR with just `skip_special_tokens=False` in the next hour.

**Comment by pramodith** [comment]
@LeonEricsson this one is ready for review now. I've mentioned some of the tradeoffs this solution entails in the description of the PR.

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -260,7 +262,10 @@ class GRPOConfig(TrainingArguments):
     max_prompt_length: Optional[int] = field(
         default=512,
         metadata={
-            "help": "Maximum length of the prompt. If the prompt is longer than this value, it will be truncated left."
+            "help": "Maximum length of the prompt. If the prompt is longer than this value, it will be truncated left. If the "
+            "prompt is conversational, we only truncate the message tokens starting from the top of the conversation
+            "and do not account for any tokens introduced
```
Comment (LeonEricsson) [inline trl/trainer/grpo_config.py]:
```suggestion
            "prompt is conversational, we only truncate the message tokens starting from the top of the conversation "
            "and do not account for any tokens introduced "
```

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1066,26 +1066,75 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
...
+        # If it is conversational the truncation preserves the chat template.
+        if not is_conversational(prompts[0]):
+            prompt_text = [x["prompt"] for x in prompts]
+            prompt_inputs = self.processing_class(
+                text=prompt_text, return_tensors="pt", padding=True, padding_side="left", add_special_tokens=False
+            )
+            prompt_inputs = super()._prepare_inputs(prompt_inputs)
+            prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
+
+            if self.max_prompt_length is not None and prompt_mask.sum(-1).max() > self.max_prompt_length:
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py:1089]: could you motivate the need for `prompt_mask.sum(-1).max() > self.max_prompt_length`? is it to avoid unnecessary decode if we don't need to truncate?

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1066,26 +1066,75 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
...
-        prompt_inputs = super()._prepare_inputs(prompt_inputs)
-        prompt_ids, prompt_mask = prompt_inputs["input_ids"], prompt_inputs["attention_mask"]
-
-        if self.max_prompt_length is not None:
-            prompt_ids = prompt_ids[:, -self.max_prompt_length :]
-            prompt_mask = prompt_mask[:, -self.max_prompt_length :]
-            prompts_text = self.processing_class.batch_decode(
-                prompt_ids, skip_special_tokens=False, clean_up_tokenization_spaces=False
-            )
+        prompt_ids, prompt_mask, prompts_text = self._get_prompt_inputs(inputs)
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]: let's make naming consistent `prompts_ids`, `prompts_mask` ...

**Review by LeonEricsson — COMMENTED** [review]
I tested running with the following conversational dataset:

```
SYSTEM_PROMPT = """
Respond in the following format:
<reasoning>
...
</reasoning>
<answer>
...
</answer>

"""
def get_gsm8k_questions(split="train") -> Dataset:
    data = load_dataset("openai/gsm8k", "main")[split]  # type: ignore
    data = data.map(
        lambda x: {  # type: ignore
            "prompt": [{"role": "system", "content": SYSTEM_PROMPT}, {"role": "user", "content": x["question"]}],
            "answer": extract_hash_answer(x["answer"]),
        }
    )  # type: ignore
    return data  # type: ignore
```

and the `prompts_text` returned by `_get_prompt_inputs` had the wrong system prompt. They contained the tokenizer's default system prompt instead.

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -1066,26 +1066,75 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
...
+                                num_tokens_to_truncate = 0
+                                truncated_messages[-1].append(msg)
+
+                prompt_inputs = self.processing_class.apply_chat_template(
+                    truncated_messages, return_dict=True, add_generation_prompt=True
+                )
+                prompt_inputs = super()._prepare_inputs(prompt_inputs)
+                prompt_ids = prompt_inputs["input_ids"]
+                prompt_mask = prompt_inputs["attention_mask"]
+                prompt_text = self.processing_class.batch_decode(prompt_ids, skip_special_tokens=False)
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]: this should be outside/after the for-loop, right?

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py:1089]: Yes exactly, gets rid of any redundant ops.

Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py:1089]: Got it. Given we've already padded, won't `prompt_mask.sum(-1).max()` always equal `prompt_ids.shape[-1]` or `prompt_mask.shape[-1]` (we're doing 'longest' padding - _pad to the longest sequence in the batch_)?

Comment (pramodith) [reply trl/trainer/grpo_trainer.py:1089]: Yeah that's fair. I'll change that to just use the shape.

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: Yep, good catch.

File: trl/trainer/grpo_trainer.py (hunk 3 above)
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: changing the naming to `prompts_ids` and `prompts_mask` means that we might also want to update the dictionary keys in the returned dict of this function. Meaning that we'll also need to update the keys used to access these tensors throughout the file, just wanted to double check if you're okay with that.

https://github.com/huggingface/trl/blob/02cce41d062a6b2754dfc2f5a31437d475003e45/trl/trainer/grpo_trainer.py#L1363-L1371

Comment (LeonEricsson) [reply trl/trainer/grpo_trainer.py]: okay, leave it as is. This inconsistency was around before this PR.

**Comment by pramodith** [comment]
> I tested running with the following conversational dataset:
> 
> ```
> SYSTEM_PROMPT = """
> Respond in the following format:
> <reasoning>
> ...
> </reasoning>
> <answer>
> ...
> </answer>
> 
> """
> def get_gsm8k_questions(split="train") -> Dataset:
>     data = load_dataset("openai/gsm8k", "main")[split]  # type: ignore
>     data = data.map(
>         lambda x: {  # type: ignore
>             "prompt": [{"role": "system", "content": SYSTEM_PROMPT}, {"role": "user", "content": x["question"]}],
>             "answer": extract_hash_answer(x["answer"]),
>         }
>     )  # type: ignore
>     return data  # type: ignore
> ```
> 
> and the `prompts_text` returned by `_get_prompt_inputs` had the wrong system prompt. They contained the tokenizer's default system prompt instead.

What was the `max_prompt_length` configured to? Some chat templates automatically introduce the default system prompt if the system message is missing in the messages for a given batch. This happens when running the `apply_chat_template` function.

So I'm assuming that the `max_prompt_length` that was set for this example eliminated the custom system prompt. Now that I think about it this means that any custom system prompt will always be pruned. Might make more sense for us to start pruning from the first user message instead of starting from the system prompt and only prune from the system prompt in case all turns barring the last user message was pruned.

But I suppose that would make this approach inconsistent with how the non-chat prompt truncation is handled. Let me know what you think the best approach is.

**Comment by LeonEricsson** [comment]
> What was the `max_prompt_length` configured to? Some chat templates automatically introduce the default system prompt if the system message is missing in the messages for a given batch. This happens when running the `apply_chat_template` function.

yes good catch, i believe this was indeed the case.

> So I'm assuming that the `max_prompt_length` that was set for this example eliminated the custom system prompt. Now that I think about it this means that any custom system prompt will always be pruned. Might make more sense for us to start pruning from the first user message instead of starting from the system prompt and only prune from the system prompt in case all turns barring the last user message was pruned.

> But I suppose that would make this approach inconsistent with how the non-chat prompt truncation is handled. Let me know what you think the best approach is.

tbh, I'm not entirely sure what the correct approach is here. I feel like user content should indeed be truncated first. However, I agree that we're moving into territory where handling conversational versus non-conversational formats significantly diverges.

**Comment by pramodith** [comment]
> tbh, I'm not entirely sure what the correct approach is here. I feel like user content should indeed be truncated first. However, I agree that we're moving into territory where handling conversational versus non-conversational formats significantly diverges.

Do we want to rope in someone else for a third opinion? I don't think there's a right approach/answer here prompt truncation is going to always be less than ideal in terms of how it affects training and one can make a case for either approach.

**Comment by pramodith** [comment]
@qgallouedec can you please give a third opinion on this PR?

**Comment by qgallouedec** [comment]
Sorry for the late reply. I'm slowly catching up after my vacation
I agree with @LeonEricsson, I understand the intuition of not breaking the chat template, but do we have solid results to validate this?:

> a naive truncation will break the template leading to poor training results.

And are there any results to show that this approach gives significantly better results? I've never seen a paper or lib using this approach, do you know of any?

**Comment by pramodith** [comment]
> And are there any results to show that this approach gives significantly better results? I've never seen a paper or lib using this approach, do you know of any?

That's a fair question, its perhaps necessary to quantify the impact of this. I can create a colab notebook to test out the results of a model trained on a dataset where 20% of the prompts are truncated. We can then compare the results on a test set when preserving and not preserving the chat template.

Are there any datasets that you'd recommend for this and what model should I be training, is Qwen3-0.6B good enough or do we want to train a larger model?

**Comment by LeonEricsson** [comment]
We're left truncating, so cutting the system prompt first and foremost. We should see this impact training, given the importance of the system prompt in common GRPO applications.   

> Are there any datasets that you'd recommend for this and what model should I be training, is Qwen3-0.6B good enough or do we want to train a larger model?

How about [open-r1/DAPO-Math-17k-Processed ](https://huggingface.co/datasets/open-r1/DAPO-Math-17k-Processed ) with Qwen3 1.7B
